### PR TITLE
fix a bug in sparse batch loader when batch size is extremely large

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
The previous implementation of sparse batch loader waits for the number of data instances reaches `batch_size` before allocating the buffer for copy (unlike dense batch loader, we don't know the actual buffer size to allocate since the number of non-zeros varies). However, such delayed allocate is buggy when batch size is extremely large, because the `DataInst` returned by `base_` is only a reference to the data. By the time copy happens, the referenced data might already be updated by the parser. 

This PR modifies the sparse batch loader to be almost the same as batch loader, except that:
- data buffer is allocated based on estimation, and adjusted when capacity is not sufficient
- before returning the output, the shape of tensor has to reflect the actual number of elements in the tensor.

@cjolivier01 @reminisce @anirudh2290  @ZiyueHuang

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/io/iter_sparse_batchloader.h [hunk 1]
```diff
@@ -157,44 +140,100 @@ class SparseBatchLoader : public BatchLoader, public SparseIIterator<TBlobBatch>
...
-    size_t total_size = inst_cache_[0].data.size();
+    size_t total_size = first_batch.data.size();
     data_.resize(total_size);
     offsets_.resize(total_size, 0);
-    std::vector<size_t> vec_sizes(total_size, 0);
-    // accumulate the memory required for a batch
+    // tensor buffer sizes
+    std::vector<size_t> buff_sizes(total_size, 0);
+    dtypes_.resize(total_size);
+    out_.data.resize(total_size);
```
Comment (cjolivier01) [inline src/io/iter_sparse_batchloader.h]: How big can total_size get?

File: src/io/iter_sparse_batchloader.h [hunk 2]
```diff
@@ -157,44 +140,100 @@ class SparseBatchLoader : public BatchLoader, public SparseIIterator<TBlobBatch>
...
+      mshadow::Copy(data_[i].get<cpu, 1, DType>().Slice(0, src_size), temp.get<cpu, 1, DType>());
+    });
+  }
+
+  /* \brief copy the data instance to data buffer */
+  void CopyData(const DataInst& d, const size_t top) {
+    int64_t unit_size = 0;
+    out_.inst_index[top] = d.index;
+    for (size_t i = 0; i < d.data.size(); ++i) {
+      if (!IsIndPtr(i)) {
```
Comment (cjolivier01) [inline src/io/iter_sparse_batchloader.h]: How big can d.data.size() get?

Comment (eric-haibin-lin) [reply src/io/iter_sparse_batchloader.h]: usually at most 6

File: src/io/iter_sparse_batchloader.h (hunk 1 above)
Comment (eric-haibin-lin) [reply src/io/iter_sparse_batchloader.h]: at most 6

**Comment by eric-haibin-lin** [comment]
Closing it for now until the test is fixed
